src/components/cuda/lcuda_dispatch.c

```diff
@@ -0,0 +1,232 @@
+/**
+ * @file    lcuda_dispatch.c
+ * @author  Anustuv Pal
+ *          anustuv@icl.utk.edu
+ */
+
+#include "lcuda_config.h"
+#include "lcuda_utils.h"
+#include "lcuda_dispatch.h"
+#include "lcuda_debug.h"
+
+#if defined(API_PERFWORKS)
+#   include "cupti_profiler.h"
+#endif
+
+#if defined(API_EVENTS)
+#   include "cupti_events.h"
+#endif
+
+void cuptid_shutdown(void)
+{
+    if (util_runtime_is_perfworks_api()) {
+
+#if defined(API_PERFWORKS)
+        cuptip_shutdown();
+#endif
+
+    } else if (util_runtime_is_events_api()) {
+
+#if defined(API_EVENTS)
+        cuptie_shutdown();
+#endif
+
+    }
+
+    util_unload_cuda_sym();
+}
+
+int cuptid_init(const char **pdisabled_reason)
+{
+    int papi_errno;
+    papi_errno = util_load_cuda_sym(pdisabled_reason);
+    if (papi_errno != PAPI_OK) {
+        goto fn_exit;
+    }
+
+    if (util_gpu_collection_kind() == GPU_COLLECTION_MIXED) {
+        *pdisabled_reason = "No support for systems with mixed compute capabilities, such as CC < 7.0 and CC > 7.0 GPUS.";
+        papi_errno = PAPI_ECMP;
+        goto fn_exit;
+    }
+
+    if (util_runtime_is_perfworks_api()) {
+
+#if defined(API_PERFWORKS)
+        papi_errno = cuptip_init(pdisabled_reason);
+#else
+        *pdisabled_reason = "PAPI not built with NVIDIA profiler API support.";
+        papi_errno = PAPI_ECMP;
+        goto fn_exit;
+#endif
+
+    } else if (util_runtime_is_events_api()) {
+
+#if defined(API_EVENTS)
+        papi_errno = cuptie_init(pdisabled_reason);
+#else
+        *pdisabled_reason = "Unknown events API problem.";
+        papi_errno = PAPI_ECMP;
+#endif
+
+    } else {
+        *pdisabled_reason = "CUDA configuration not supported.";
+        papi_errno = PAPI_ECMP;
+    }
+fn_exit:
+    return papi_errno;
+}
+
+int cuptid_thread_info_init(void **thread_info)
```

Comment:
I would also say replace init with create and add an update function as well.

Reply:
I dont think `attr_init` is a good name for this. Cuda contexts are threading specific functionality. If we dont want cuda context in the name, thread_info is the best alternative.

Reply:
My point is this is control related information. Control information is thread specific.

Reply:
Sure control state is already thread specific, but why can it not possess information related to that particular thread? Is there any limitation on that?

Reply:
It possesses information about a particular thread like the control state in the frontend does. The point is: making explicit references to thread is superfluous and confusing because the backend deals with `control` (which is thread specific). I think using `thread` in the name is also inconsistent. If we pass that information to the control create function it makes sense to call the info object as `control_info` rather than `thread_info`.